src/jsonrpc.c

```diff
@@ -0,0 +1,493 @@
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <stdbool.h>
+
+#include <yajl/yajl_tree.h>
+#include <yajl/yajl_gen.h>
+
+#include "kore.h"
+#include "http.h"
+#include "jsonrpc.h"
+
+static void
+init_log(struct jsonrpc_log *log)
+{
+	log->msg = NULL;
+	log->next = log;
+	log->prev = log;
+}
+
+static void
+append_log(struct jsonrpc_log *prev, int lvl, char *msg)
+{
+	struct jsonrpc_log *new = kore_malloc(sizeof(struct jsonrpc_log));
+
+	new->lvl = lvl;
+	new->msg = msg;
+
+	new->prev = prev;
+	new->next = prev->next;
+	prev->next->prev = new;
+	prev->next = new;
+}
+
+static void
+free_log(struct jsonrpc_log *root)
+{
+	for (struct jsonrpc_log *it = root->next; it != root; it = it->next) {
+		kore_mem_free(it);
+	}
+}
+
+static void
+init_request(struct jsonrpc_request *req)
+{
+	init_log(&req->log);
+	kore_buf_init(&req->buf, 256);
+	req->gen = NULL;
+	req->http = NULL;
+	req->json = NULL;
+	req->id = NULL;
+	req->method = NULL;
+	req->params = NULL;
+	req->log_levels = (1 << LOG_EMERG) | (1 << LOG_ERR) | (1 << LOG_WARNING)
+			| (1 << LOG_NOTICE);
+        req->flags = 0;
+}
+
+static void
+free_request(struct jsonrpc_request *req)
+{
+	if (req->gen != NULL) {
+		yajl_gen_free(req->gen);
+		req->gen = NULL;
+	}
+	if (req->json != NULL) {
+		yajl_tree_free(req->json);
+		req->json = NULL;
+	}
+	kore_buf_destroy(&req->buf);
+	free_log(&req->log);
+}
+
+void
+jsonrpc_log(struct jsonrpc_request *req, int lvl, const char *fmt, ...)
+{
+	va_list		ap;
+	char		*msg;
+	u_int32_t	start = req->buf.offset;
+
+	va_start(ap, fmt);
+	kore_buf_appendv(&req->buf, fmt, ap);
+	va_end(ap);
+
+	msg = kore_buf_stringify(&req->buf, NULL) + start;
+
+	append_log(&req->log, lvl, msg);
+}
+
+static int
+read_json_body(struct http_request *http_req, ssize_t body_max_len,
+    struct jsonrpc_request *req)
+{
+	char		*body_string;
+	u_int32_t	body_start = req->buf.offset;
+	ssize_t		body_len = 0, chunk_len;
+	u_int8_t	chunk_buffer[BUFSIZ];
+	char		error_buffer[1024];
+
+	for (;;) {
+		chunk_len = http_body_read(http_req, chunk_buffer,
+			sizeof(chunk_buffer));
+		if (chunk_len == -1) {
+			jsonrpc_log(req, LOG_CRIT,
+			    "Failed to read request body");
+			return (JSONRPC_SERVER_ERROR);
+		}
+
+		if (chunk_len == 0)
+			break;
+
+		if (body_len > SSIZE_MAX - chunk_len) {
+			jsonrpc_log(req, LOG_CRIT,
+			    "Request body bigger than the platform accepts");
+			return (JSONRPC_SERVER_ERROR);
+		}
+		body_len += chunk_len;
+
+		if (body_len > body_max_len) {
+			jsonrpc_log(req, LOG_ERR,
+			    "Request overreached configured body size limit");
+			return (JSONRPC_LIMIT_REACHED);
+		}
+		kore_buf_append(&req->buf, chunk_buffer, chunk_len);
+	}
+
+	/* Grab our body data as a NUL-terminated string. */
+	body_string = kore_buf_stringify(&req->buf, NULL) + body_start;
```

Comment:
It should yes always be 0, for some reason, at the time of writing, I didn't want to rely on it.
I'll hardcode it.